File: ats_software/Threads/Src/calculate.c
#include "calculate.h"

void calculate_update(struct GpsData* gps_data, struct TargetPosition* target_position) {
	// copy out gps data
	osMutexAcquire(gps_data->mtx, 0);
	double self_lat = TO_RADIANS(gps_data->self_lat);
	double self_lon = TO_RADIANS(gps_data->self_lon);
	double self_alt = gps_data->self_alt;
	double target_lat = TO_RADIANS(gps_data->target_lat);
	double target_lon = TO_RADIANS(gps_data->target_lon);
	double target_alt = gps_data->target_alt;
	osMutexRelease(gps_data->mtx);

	double delta_lon = target_lon-self_lon;
	double delta_lat = target_lat-self_lat;
	double delta_alt = target_alt-self_alt;

	// calculate rotation angle from north
	double rotation_angle = TO_DEGREES(atan2(sin(delta_lon)*cos(target_lat), cos(self_lat)*sin(target_lat) - sin(self_lat)*cos(target_lat)*cos(delta_lon)));

	// calculate elevation angle from horizontal
	double a = sin(delta_lat/2)*sin(delta_lat/2) + sin(delta_lon/2)*sin(delta_lon/2)*cos(self_lat)*cos(target_lat); 
  	double c = 2*atan2(sqrt(a), sqrt(1-a)); 
	double elevation_angle = TO_DEGREES(atan2(delta_alt, EARTH_RADIUS_M*c));

	// clip elevation angle to [0, 90]
	elevation_angle = elevation_angle > 90 ? 90 : elevation_angle;
	elevation_angle = elevation_angle < 0 ? 0 : elevation_angle;

	// make rotation angle always positive
	rotation_angle = rotation_angle < 0 ? rotation_angle + 360 : rotation_angle;

	// write target position data
	osMutexAcquire(target_position->mtx, 0);
	target_position->rotation_angle = rotation_angle;
	target_position->elevation_angle = elevation_angle;
	osMutexRelease(target_position->mtx);

	return;
}

File: thread_tests/calculate.py
import math

EARTH_RADIUS_M = 6378137

def to_radians(val):
    return val*math.pi/180

def to_degrees(val):
    return val*180/math.pi

def compute_angles(self_gps, target_gps):
    self_lat = to_radians(self_gps[0])
    self_lon = to_radians(self_gps[1])
    self_alt = self_gps[2]
    target_lat = to_radians(target_gps[0])
    target_lon = to_radians(target_gps[1])
    target_alt = target_gps[2]

    delta_lon = target_lon-self_lon
    delta_lat = target_lat-self_lat
    delta_alt = target_alt-self_alt

    rotation_angle = to_degrees(math.atan2(math.sin(delta_lon)*math.cos(target_lat), math.cos(self_lat)*math.sin(target_lat) - math.sin(self_lat)*math.cos(target_lat)*math.cos(delta_lon)))

    a = math.sin(delta_lat/2)*math.sin(delta_lat/2) + math.sin(delta_lon/2)*math.sin(delta_lon/2)*math.cos(self_lat)*math.cos(target_lat)
    c = 2*math.atan2(math.sqrt(a), math.sqrt(1-a))
    elevation_angle = to_degrees(math.atan2(delta_alt, EARTH_RADIUS_M*c))

    if rotation_angle < 0:
        rotation_angle += 360

    return rotation_angle, elevation_angle

def main():
    self_gps = (38.27269,-84.94623, 0)
    target_gps = (34.5591396, -87.9065571, 1000)

    rotation, elevation = compute_angles(self_gps, target_gps)
    print(f"rotation angle: {rotation}  elevation angle: {elevation}")

if __name__ == "__main__":
    main()